### PR TITLE
Expose deletion API for projects/features

### LINKED ISSUE #842: [FR] Expose the deletion API for features/projects
### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### Feature Request Proposal

Expose the deletion API for features/projects

### Motivation

> #### What is the use case for this feature?
Expose the deletion API for features/projects

### Details

_No response_

### What component(s) does this feature request affect?

- [ ] `Python Client`: This is the client users use to interact with most of our API. Mostly written in Python.
- [ ] `Computation Engine`: The computation engine that execute the actual feature join and generation work. Mostly in Scala and Spark.
- [ ] `Feature Registry API`: The frontend API layer supports SQL, Purview(Atlas) as storage. The API layer is in Python(FAST API)
- [ ] `Feature Registry Web UI`: The Web UI for feature registry. Written in React

### PR DESCRIPTION
**Pull request by aabbasi-hbo**
## Description

Resolves #842 

Exposes the deletion API for features/projects

## How was this PR tested?

Updated unit tests

## Does this PR introduce any user-facing changes?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure to clarify your proposed changes.

### CODE REVIEW

**Review by aabbasi-hbo — COMMENTED** [review]
@xiaoyongzhu is there a reason why the registry purview client (https://github.com/feathr-ai/feathr/blob/main/feathr_project/feathr/registry/_feature_registry_purview.py) does not call the API specified in (https://github.com/feathr-ai/feathr/blob/main/registry/purview-registry/main.py)? It seems that API functionality has been directly implemented in the client

**Comment by aabbasi-hbo** [comment]
> @xiaoyongzhu is there a reason why the registry purview client (https://github.com/feathr-ai/feathr/blob/main/feathr_project/feathr/registry/_feature_registry_purview.py) does not call the API specified in (https://github.com/feathr-ai/feathr/blob/main/registry/purview-registry/main.py)? It seems that API functionality has been directly implemented in the client

Nevermind I see that it is deprecated

**Comment by xiaoyongzhu** [comment]
Thanks @aabbasi-hbo for the PR! This is very helpful. 

@windoze @YihuiGuo can you take a look as well?

**Comment by windoze** [comment]
Cascading deletion seems to be a too-dangerous operation, could we have some safer alternatives?
My idea is to add another method to retrieve all downstream, so we can know the correct order to delete entity one by one

**Comment by aabbasi-hbo** [comment]
> 



> Cascading deletion seems to be a too-dangerous operation, could we have some safer alternatives? My idea is to add another method to retrieve all downstream, so we can know the correct order to delete entity one by one

@windoze which cascading operation are you referring to? The deletion uses the BFS function to get all downstream entities and deletes them through another BFS

**Comment by windoze** [comment]
'Cascading" means one delete request could actually delete a lot of entities without any prior notice, which could be too dangerous.
I prefer to add 2 APIs, one to get a list of entities that depend on the specific one in the correct order, and another one to delete one entity at a time, refusing to delete an entity that has any downstream.

**Comment by aabbasi-hbo** [comment]
> 'Cascading" means one delete request could actually delete a lot of entities without any prior notice, which could be too dangerous. I prefer to add 2 APIs, one to get a list of entities that depend on the specific one in the correct order, and another one to delete one entity at a time, refusing to delete an entity that has any downstream.

got it. yeah that makes sense - will update. Thanks for the feedback!

**Review by aabbasi-hbo — COMMENTED** [review]
@windoze mind reviewing again when you get a chance? thanks!

File: registry/purview-registry/main.py [hunk 1]
```diff
@@ -56,6 +56,22 @@ def get_projects_ids() -> dict:
...
+    downstream_entities = registry.get_dependent_entities(entity_id)
+    return list([e.to_dict() for e in downstream_entities])
+
+@router.delete("/entity/{entity}")
+def delete_entity(entity: str):
+    entity_id = registry.get_entity_id(entity)
+    downstream_entities = registry.get_dependent_entities(entity_id)
+    if len(downstream_entities) > 0:
+        raise HTTPException(
+           status_code=500, detail=f"""Entity cannot be deleted as it has downstream/dependent entities.
```
Comment (windoze) [inline registry/purview-registry/main.py:71]: Better to use a status code other than 500, my recommendation is "412 precondition failed".

File: registry/sql-registry/main.py [hunk 2]
```diff
@@ -86,6 +86,22 @@ def get_projects_ids() -> dict:
...
+    downstream_entities = registry.get_dependent_entities(entity_id)
+    return list([e.to_dict() for e in downstream_entities])
+
+@router.delete("/entity/{entity}")
+def delete_entity(entity: str):
+    entity_id = registry.get_entity_id(entity)
+    downstream_entities = registry.get_dependent_entities(entity_id)
+    if len(downstream_entities) > 0:
+        raise HTTPException(
+            status_code=500, detail=f"""Entity cannot be deleted as it has downstream/dependent entities.
```
Comment (windoze) [inline registry/sql-registry/main.py:101]: Better to use a status code other than 500, my recommendation is "412 precondition failed".

**Review by windoze — COMMENTED** [review]
It looks good to me, thanks for a great job.

**Comment by windoze** [comment]
> @windoze mind reviewing again when you get a chance? thanks!

Sorry for the delay, I was working on something else.
I've put my comments, please take a look.

Thanks for the contribution!

File: registry/purview-registry/main.py (hunk 1 above)
Comment (aabbasi-hbo) [reply registry/purview-registry/main.py]: updated

File: registry/sql-registry/main.py (hunk 2 above)
Comment (aabbasi-hbo) [reply registry/sql-registry/main.py]: updated

**Review by xiaoyongzhu — APPROVED** [review]
This is a really nice PR, thanks for the contribution!

**Review by windoze — APPROVED** [review]

**Pull request merged**